feat: LP Asset rebalancing (PRO-1259)

Janislav opened the pull request:
# Pull Request

Closes: PRO-1259

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have updated documentation where appropriate.

## Summary

This PR adds an extrinsic that allows LPs to move assets between accounts without the need to withdrawal first. The existing extrinsic and LP-API endpoints doesn't change, which should make it none-breaking for the LPs. The changes in the LP-pallet could affect the engine, though, which make it breaking. I think we have to check this.

#### Non-Breaking changes

If this PR includes non-breaking changes, select the `non-breaking` label. On merge, CI will automatically cherry-pick the commit to a PR against the release branch.

Janislav commented:
For now, there is only a unit-test. If we are happy with the PR, I would also add a bouncer test.

dandanlen reviewed (commented):
Looking good - will take a closer look on Monday. Only comment so far is that I think we should come up with a better name than `move_asset`. Will think about it....

dandanlen commented:
I took another look, changed the naming to 'transfer' instead of 'move'.

As discussed, I think a bouncer test would be useful for regression prevention.

dandanlen commented:
Apart from that this is ready to go. Nice work.

dandanlen commented on api/lib/src/lp.rs:
```diff
@@ -256,6 +256,29 @@ pub trait LpApi: SignedExtrinsicApi + Sized + Send + Sync + 'static {
+			.await
+			.until_in_block()
+			.await
+			.context("Registration for Liquidity Refund Address failed.")?;
```
Just noticed this:

```suggestion
			.context("Unable to transfer asset.")?;
```

dandanlen commented on state-chain/pallets/cf-lp/src/lib.rs:
```diff
@@ -347,6 +349,10 @@ impl<T: Config> Pallet<T> {
 				AccountOrAddress::Internal(destination_account) => {
+					ensure!(
+						account_id != destination_account,
+						Error::<T>::CannotTransferToOriginAccount
```
Good idea 👍

Janislav commented:
Docs are updated here: https://github.com/chainflip-io/chainflip-docs/pull/251

Janislav commented:
I think this should be ready now. I added some more things:

- Ensure that destination_account != source_account
- Bouncer test that is testing a balance transfer between LP accounts
- Updated the docs
- Changed amount parameter to hex string only

dandanlen commented on bouncer/shared/lp_api_test.ts:
```diff
@@ -167,11 +175,17 @@ async function testTransferAsset() {
+  const newBalanceDestination = await getLpBalance(destinationLpAccount.address);
 
-  assert(newBalance > oldBalance, `Failed to observe balance increase after transfer`);
+  assert(
+    newBalanceDestination > oldBalanceDestination,
+    `Failed to observe balance increase after transfer for destination account!`,
+  );
+  assert(
+    newBalancesSource < oldBalanceSource,
+    `Failed to observe balance decrease after transfer for source account!`,
+  );
```
Should the balances be equal before and after?

Ie. can we improve this check even further and test for `oldBalanceSource + oldBalanceDestination === newBalancesSource + newBalanceDestination` ?

Janislav commented:
I guess this is ready.
